```
RemoteFileOutboundGatewayTests > testMoveWithMkDirs FAILED
    org.springframework.messaging.MessagingException at RemoteFileOutboundGatewayTests.java:304
        Caused by: java.io.IOException at RemoteFileOutboundGatewayTests.java:304
```

--- a/spring-integration-file/src/main/java/org/springframework/integration/file/remote/gateway/AbstractRemoteFileOutboundGateway.java
+++ b/spring-integration-file/src/main/java/org/springframework/integration/file/remote/gateway/AbstractRemoteFileOutboundGateway.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2002-2019 the original author or authors.
+ * Copyright 2002-2020 the original author or authors.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
@@ -121,6 +121,7 @@ public abstract class AbstractRemoteFileOutboundGateway<F> extends AbstractReply
 			MessageSessionCallback<F, ?> messageSessionCallback) {
 
 		this(new RemoteFileTemplate<>(sessionFactory), messageSessionCallback);
+		remoteFileTemplateExplicitlySet(false);
 	}
 
 	/**
@@ -165,6 +166,7 @@ public abstract class AbstractRemoteFileOutboundGateway<F> extends AbstractReply
 			@Nullable String expression) {
 
 		this(new RemoteFileTemplate<>(sessionFactory), command, expression);
+		remoteFileTemplateExplicitlySet(false);
 	}
 
 	/**
@@ -516,6 +518,9 @@ public abstract class AbstractRemoteFileOutboundGateway<F> extends AbstractReply
 		}
 
 		populateBeanFactoryIntoComponentsIfAny();
+		if (!this.remoteFileTemplateExplicitlySet) {
+			this.remoteFileTemplate.afterPropertiesSet();
+		}
 	}
 
 	private void populateBeanFactoryIntoComponentsIfAny() {

--- a/spring-integration-file/src/test/java/org/springframework/integration/file/remote/gateway/RemoteFileOutboundGatewayTests.java
+++ b/spring-integration-file/src/test/java/org/springframework/integration/file/remote/gateway/RemoteFileOutboundGatewayTests.java
@@ -295,7 +295,7 @@ public class RemoteFileOutboundGatewayTests {
 		final List<String> madeDirs = new ArrayList<>();
 		doAnswer(invocation -> {
 			madeDirs.add(invocation.getArgument(0));
-			return null;
+			return true;
 		}).when(session).mkdir(anyString());
 		when(sessionFactory.getSession()).thenReturn(session);
 		Message<String> requestMessage = MessageBuilder.withPayload("foo")
@@ -939,7 +939,6 @@ public class RemoteFileOutboundGatewayTests {
 		Session<TestLsEntry> session = mock(Session.class);
 		TestRemoteFileOutboundGateway gw = new TestRemoteFileOutboundGateway(sessionFactory, "mput", "payload");
 		gw.setRemoteDirectoryExpression(new LiteralExpression("foo/"));
-		gw.setBeanFactory(mock(BeanFactory.class));
 		gw.afterPropertiesSet();
 		when(sessionFactory.getSession()).thenReturn(session);
 		final AtomicReference<String> written = new AtomicReference<>();

--- a/spring-integration-sftp/src/test/java/org/springframework/integration/sftp/outbound/SftpOutboundTests.java
+++ b/spring-integration-sftp/src/test/java/org/springframework/integration/sftp/outbound/SftpOutboundTests.java
@@ -219,7 +219,7 @@ public class SftpOutboundTests {
 		final List<String> madeDirs = new ArrayList<>();
 		doAnswer(invocation -> {
 			madeDirs.add(invocation.getArgument(0));
-			return null;
+			return true;
 		}).when(session).mkdir(anyString());
 		handler.handleMessage(new GenericMessage<>("qux"));
 		assertThat(madeDirs.size()).isEqualTo(3);

